File: src/zlac706_driver/zlac706_driver/zlac706_diffdrive_node.py
#!/usr/bin/env python3
"""
ZLAC706-RC 差速輪 ROS2 節點 (rclpy)
------------------------------------------------
底層 Modbus RTU 通訊採用手動組封包 + CRC16（不依賴 pymodbus），
協定與暫存器對應到已驗證可動的測試腳本：
  - 左輪 slave address = 0x01, 右輪 slave address = 0x02
  - 寫入速度命令：暫存器 0x0011 (function code 0x06)
  - 讀取狀態+速度：暫存器 0x00D2, 長度 2 (function code 0x03)
  - 使能指令：對 0x0010 寫入固定值（每輪各自的 CRC 已內嵌）
  - RPM <-> 內部整數換算：data = rpm / 3000 * 8192

安裝需求 (在 Jetson Nano 上執行)：
  pip3 install pyserial --break-system-packages
  sudo usermod -a -G dialout $USER   # 需要重新登入才生效

功能：
  - 訂閱 /cmd_vel (geometry_msgs/Twist)，換算成左右輪目標 RPM，寫入驅動板
  - 定時讀取左右輪實際 RPM，推算里程計，發布 /odom 與 TF (odom -> base_link)
  - 逾時未收到新的 /cmd_vel 會自動歸零（安全機制）
"""

import time
import math

import serial

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist, TransformStamped
from nav_msgs.msg import Odometry
from tf2_ros import TransformBroadcaster


# ---------- Modbus CRC ----------
def modbus_crc(arr: bytes) -> int:
    crc = 0xFFFF
    for b in arr:
        crc ^= b
        for _ in range(8):
            if crc & 1:
                crc = (crc >> 1) ^ 0xA001
            else:
                crc >>= 1
    return crc & 0xFFFF


def append_crc(payload: bytearray) -> bytes:
    crc = modbus_crc(payload)
    payload += bytes([crc & 0xFF, (crc >> 8) & 0xFF])  # little-endian CRC
    return bytes(payload)


def build_speed_cmd(addr: int, rpm: float) -> bytes:
    """寫單一保持暫存器 0x0011：速度命令 (data = rpm/3000 * 8192)"""
    data = int(rpm / 3000.0 * 8192.0)
    data &= 0xFFFF  # 兩補數
    payload = bytearray([addr, 0x06, 0x00, 0x11, (data >> 8) & 0xFF, data & 0xFF])
    return append_crc(payload)


def read_strpm_cmd(addr: int) -> bytes:
    """讀保持暫存器 0x00D2, 長度 2（狀態 + 速度）"""
    payload = bytearray([addr, 0x03, 0x00, 0xD2, 0x00, 0x02])
    return append_crc(payload)


def parse_speed_reply(data: bytes, expect_addr: int):
    """回傳 (state:int, speed_raw:int16)；格式不對回傳 (None, None)"""
    if len(data) >= 9 and data[0] == expect_addr and data[1] == 0x03 and data[2] == 0x04:
        state = (data[3] << 8) | data[4]
        raw = (data[5] << 8) | data[6]
        if raw >= 0x8000:
            raw -= 0x10000
        return state, raw
    return None, None


def rpm_from_raw(raw: int) -> float:
    return raw * 3000.0 / 8192.0


# 使能命令（沿用已驗證可動的固定 bytes，內含各自的 CRC）
ENABLE_L = bytes([0x01, 0x06, 0x00, 0x10, 0x00, 0x1F, 0xC9, 0xC7])
ENABLE_R = bytes([0x02, 0x06, 0x00, 0x10, 0x00, 0x1F, 0xC9, 0xF4])

ADDR_L = 0x01
ADDR_R = 0x02

# 0x06 的回應是原封包回音 8 bytes；0x03 讀 2 個暫存器是 9 bytes
WRITE_REPLY_LEN = 8
READ_REPLY_LEN = 9
REPLY_TIMEOUT = 0.03
# Modbus RTU 幀間靜默：38400bps 下 3.5 字元約 0.9ms，取 2ms 保守值
SILENT_INTERVAL = 0.002


class ZLAC706DiffDrive(Node):
    def __init__(self):
        super().__init__('zlac706_diffdrive_node')

        # ---- 參數，依實際機構調整 ----
        self.declare_parameter('port', '/dev/ttyUSB0')
        self.declare_parameter('baudrate', 38400)
        self.declare_parameter('wheel_radius', 0.0508)   # 公尺
        self.declare_parameter('wheel_separation', 0.23) # 公尺，左右輪中心距
        self.declare_parameter('max_rpm', 200.0)
        self.declare_parameter('publish_rate_hz', 10.0)
        self.declare_parameter('invert_left', False)
        self.declare_parameter('invert_right', True)  # 左右輪馬達通常反向安裝
        self.declare_parameter('cmd_timeout', 0.5)

        port = self.get_parameter('port').value
        baud = self.get_parameter('baudrate').value
        self.wheel_radius = self.get_parameter('wheel_radius').value
        self.wheel_sep = self.get_parameter('wheel_separation').value
        self.max_rpm = self.get_parameter('max_rpm').value
        self.invert_left = self.get_parameter('invert_left').value
        self.invert_right = self.get_parameter('invert_right').value
        self.cmd_timeout = self.get_parameter('cmd_timeout').value

        # ---- 序列埠連線 ----
        try:
            self.ser = serial.Serial(port, baud, timeout=REPLY_TIMEOUT)
            self.get_logger().info(f'序列埠已開啟: {port} @ {baud}bps')
            self._enable_driver()
        except Exception as e:
            self.ser = None
            self.get_logger().error(f'無法開啟序列埠 {port}: {e}')

        # ---- ROS interfaces ----
        self.cmd_sub = self.create_subscription(Twist, '/cmd_vel', self.cmd_vel_cb, 10)
        self.odom_pub = self.create_publisher(Odometry, '/odom', 10)
        self.tf_broadcaster = TransformBroadcaster(self)

        # ---- 里程計狀態 ----
        self.x = 0.0
        self.y = 0.0
        self.theta = 0.0
        self.last_time = self.get_clock().now()

        # 序列埠只由 control_cb 這一條路徑存取，cmd_vel 僅更新目標值，避免搶匯流排
        self.target_rpm_l = 0.0
        self.target_rpm_r = 0.0
        self.last_cmd_time = time.time()

        period = 1.0 / self.get_parameter('publish_rate_hz').value
        self.timer = self.create_timer(period, self.control_cb)

    # ----------------------------------------------------------------
    def _transact(self, request: bytes, expect_len: int) -> bytes:
        """送出一筆 Modbus RTU 請求並收完整個回應，結束後保留幀間靜默。"""
        self.ser.reset_input_buffer()
        self.ser.write(request)
        self.ser.flush()
        data = self.ser.read(expect_len)
        time.sleep(SILENT_INTERVAL)
        return data

    def _enable_driver(self):
        if self.ser is None:
            return
        self._transact(ENABLE_L, WRITE_REPLY_LEN)
        self._transact(ENABLE_R, WRITE_REPLY_LEN)

    def _write_wheel_rpm(self, left_rpm: float, right_rpm: float):
        if self.ser is None:
            return
        left_rpm = max(-self.max_rpm, min(self.max_rpm, left_rpm))
        right_rpm = max(-self.max_rpm, min(self.max_rpm, right_rpm))

        if self.invert_left:
            left_rpm = -left_rpm
        if self.invert_right:
            right_rpm = -right_rpm

        try:
            self._transact(build_speed_cmd(ADDR_L, left_rpm), WRITE_REPLY_LEN)
            self._transact(build_speed_cmd(ADDR_R, right_rpm), WRITE_REPLY_LEN)
        except Exception as e:
            self.get_logger().warn(f'寫入速度命令失敗: {e}')

    def _read_wheel_rpm(self):
        if self.ser is None:
            return None
        try:
            data_l = self._transact(read_strpm_cmd(ADDR_L), READ_REPLY_LEN)
            data_r = self._transact(read_strpm_cmd(ADDR_R), READ_REPLY_LEN)

            _, raw_l = parse_speed_reply(data_l, ADDR_L)
            _, raw_r = parse_speed_reply(data_r, ADDR_R)
            if raw_r is None or raw_l is None:
                return None

            rpm_l = rpm_from_raw(raw_l)
            rpm_r = rpm_from_raw(raw_r)
            if self.invert_left:
                rpm_l = -rpm_l
            if self.invert_right:
                rpm_r = -rpm_r
            return rpm_l, rpm_r
        except Exception as e:
            self.get_logger().warn(f'讀取轉速回授失敗: {e}')
            return None

    # ----------------------------------------------------------------
    def cmd_vel_cb(self, msg: Twist):
        self.last_cmd_time = time.time()
        # 實測發現前後方向與 /cmd_vel 定義相反，這裡整體翻轉線速度；
        # 不動 w，也不動 invert_left/invert_right，避免動到已經調對的轉向。
        v = -msg.linear.x      # m/s
        w = msg.angular.z      # rad/s

        v_left = v - (w * self.wheel_sep / 2.0)
        v_right = v + (w * self.wheel_sep / 2.0)

        self.target_rpm_l = (v_left / (2 * math.pi * self.wheel_radius)) * 60.0
        self.target_rpm_r = (v_right / (2 * math.pi * self.wheel_radius)) * 60.0

    # ----------------------------------------------------------------
    def control_cb(self):
        if time.time() - self.last_cmd_time > self.cmd_timeout:
            self.target_rpm_l = 0.0
            self.target_rpm_r = 0.0

        self._write_wheel_rpm(self.target_rpm_l, self.target_rpm_r)
        self._update_odom(self._read_wheel_rpm())

    def _update_odom(self, fb):
        now = self.get_clock().now()
        dt = (now - self.last_time).nanoseconds / 1e9
        self.last_time = now
        if fb is None or dt <= 0.0:
            return

        rpm_left, rpm_right = fb
        v_left = (rpm_left / 60.0) * 2 * math.pi * self.wheel_radius
        v_right = (rpm_right / 60.0) * 2 * math.pi * self.wheel_radius

        v = (v_left + v_right) / 2.0
        w = (v_right - v_left) / self.wheel_sep

        delta_x = v * math.cos(self.theta) * dt
        delta_y = v * math.sin(self.theta) * dt
        delta_theta = w * dt

        self.x += delta_x
        self.y += delta_y
        self.theta += delta_theta

        quat_z = math.sin(self.theta / 2.0)
        quat_w = math.cos(self.theta / 2.0)

        odom = Odometry()
        odom.header.stamp = now.to_msg()
        odom.header.frame_id = 'odom'
        odom.child_frame_id = 'base_link'
        odom.pose.pose.position.x = self.x
        odom.pose.pose.position.y = self.y
        odom.pose.pose.orientation.z = quat_z
        odom.pose.pose.orientation.w = quat_w
        odom.twist.twist.linear.x = v
        odom.twist.twist.angular.z = w
        self.odom_pub.publish(odom)

        t = TransformStamped()
        t.header.stamp = now.to_msg()
        t.header.frame_id = 'odom'
        t.child_frame_id = 'base_link'
        t.transform.translation.x = self.x
        t.transform.translation.y = self.y
        t.transform.rotation.z = quat_z
        t.transform.rotation.w = quat_w
        self.tf_broadcaster.sendTransform(t)

    def destroy_node(self):
        try:
            self._write_wheel_rpm(0.0, 0.0)
            if self.ser is not None:
                self.ser.close()
        except Exception:
            pass
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = ZLAC706DiffDrive()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/zlac706_driver_cpp/src/zlac706_diffdrive_node.cpp
// ZLAC706-RC 差速輪 ROS 2 節點 (rclcpp)
//
// Modbus RTU 手動組封包 + CRC16，協定對應已驗證可動的測試腳本：
//   - 實接線：左輪 slave address = 0x02, 右輪 slave address = 0x01
//   - 寫入速度命令：暫存器 0x0011 (function code 0x06)
//   - 讀取狀態+速度：暫存器 0x00D2, 長度 2 (function code 0x03)
//   - 使能指令：對 0x0010 寫入固定值（每輪各自的 CRC 已內嵌）
//   - RPM <-> 內部整數換算：data = rpm / 3000 * 8192
//
// 所有序列埠存取集中在單一 timer callback，cmd_vel 僅更新目標值，避免搶匯流排。

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <memory>
#include <optional>
#include <string>
#include <thread>
#include <utility>
#include <vector>

#include <errno.h>
#include <fcntl.h>
#include <sys/select.h>
#include <termios.h>
#include <unistd.h>

#include <geometry_msgs/msg/transform_stamped.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <rclcpp/rclcpp.hpp>
#include <tf2_ros/transform_broadcaster.h>

namespace
{

// 實測左右輪接在與直覺相反的 slave address，接反時只有轉向會錯、直行看不出來。
constexpr uint8_t ADDR_L = 0x02;
constexpr uint8_t ADDR_R = 0x01;

// 驅動器實測需要每幀間隔 >=30ms 才會回應，間隔由 timer 週期提供，不在 callback 內 sleep。
constexpr size_t WRITE_REPLY_LEN = 8;
constexpr size_t READ_REPLY_LEN = 9;

// 使能命令（沿用已驗證可動的固定 bytes，內含各自的 CRC）——依 address 命名，與左右無關
const std::vector<uint8_t> ENABLE_ADDR_01{0x01, 0x06, 0x00, 0x10, 0x00, 0x1F, 0xC9, 0xC7};
const std::vector<uint8_t> ENABLE_ADDR_02{0x02, 0x06, 0x00, 0x10, 0x00, 0x1F, 0xC9, 0xF4};

uint16_t modbusCrc(const std::vector<uint8_t> & data)
{
  uint16_t crc = 0xFFFF;
  for (uint8_t b : data) {
    crc ^= b;
    for (int i = 0; i < 8; ++i) {
      crc = (crc & 1) ? static_cast<uint16_t>((crc >> 1) ^ 0xA001) : static_cast<uint16_t>(crc >> 1);
    }
  }
  return crc;
}

std::vector<uint8_t> appendCrc(std::vector<uint8_t> payload)
{
  const uint16_t crc = modbusCrc(payload);
  payload.push_back(static_cast<uint8_t>(crc & 0xFF));
  payload.push_back(static_cast<uint8_t>((crc >> 8) & 0xFF));
  return payload;
}

std::vector<uint8_t> buildSpeedCmd(uint8_t addr, double rpm)
{
  const auto value = static_cast<uint16_t>(static_cast<int32_t>(rpm / 3000.0 * 8192.0) & 0xFFFF);
  return appendCrc({addr, 0x06, 0x00, 0x11,
      static_cast<uint8_t>((value >> 8) & 0xFF), static_cast<uint8_t>(value & 0xFF)});
}

std::vector<uint8_t> readStrpmCmd(uint8_t addr)
{
  return appendCrc({addr, 0x03, 0x00, 0xD2, 0x00, 0x02});
}

/// 在位元組流中找出 CRC 正確的目標幀，藉此跳過 RS-485 回音與前導雜訊。
std::optional<std::vector<uint8_t>> findFrame(
  const std::vector<uint8_t> & buffer, uint8_t addr, uint8_t func, size_t len)
{
  for (size_t i = 0; i + len <= buffer.size(); ++i) {
    if (buffer[i] != addr || buffer[i + 1] != func) {
      continue;
    }
    const std::vector<uint8_t> body(buffer.begin() + i, buffer.begin() + i + len - 2);
    const uint16_t crc = modbusCrc(body);
    if ((crc & 0xFF) == buffer[i + len - 2] && ((crc >> 8) & 0xFF) == buffer[i + len - 1]) {
      return std::vector<uint8_t>(buffer.begin() + i, buffer.begin() + i + len);
    }
  }
  return std::nullopt;
}

std::optional<int16_t> parseSpeedReply(const std::vector<uint8_t> & frame)
{
  if (frame.size() < READ_REPLY_LEN || frame[2] != 0x04) {
    return std::nullopt;
  }
  return static_cast<int16_t>((frame[5] << 8) | frame[6]);
}

std::string toHex(const std::vector<uint8_t> & data)
{
  static const char * kDigits = "0123456789ABCDEF";
  std::string out;
  out.reserve(data.size() * 3);
  for (uint8_t b : data) {
    out.push_back(kDigits[b >> 4]);
    out.push_back(kDigits[b & 0x0F]);
    out.push_back(' ');
  }
  return out;
}

double rpmFromRaw(int16_t raw)
{
  return raw * 3000.0 / 8192.0;
}

std::optional<speed_t> toBaudConstant(int baud)
{
  switch (baud) {
    case 9600: return B9600;
    case 19200: return B19200;
    case 38400: return B38400;
    case 57600: return B57600;
    case 115200: return B115200;
    case 230400: return B230400;
    default: return std::nullopt;
  }
}

/// 以 termios 操作的 RS-485 序列埠，每次交易 = 送出請求 → 收到 CRC 正確的目標幀。
class SerialPort
{
public:
  ~SerialPort() {close();}

  bool open(const std::string & device, int baud, double reply_timeout, std::string & error)
  {
    reply_timeout_ = reply_timeout;
    const auto speed = toBaudConstant(baud);
    if (!speed) {
      error = "不支援的 baudrate: " + std::to_string(baud);
      return false;
    }

    fd_ = ::open(device.c_str(), O_RDWR | O_NOCTTY | O_NONBLOCK);
    if (fd_ < 0) {
      error = device + ": " + std::strerror(errno);
      return false;
    }

    termios tty{};
    if (::tcgetattr(fd_, &tty) != 0) {
      error = std::strerror(errno);
      close();
      return false;
    }

    ::cfmakeraw(&tty);
    ::cfsetispeed(&tty, *speed);
    ::cfsetospeed(&tty, *speed);
    tty.c_cflag |= (CLOCAL | CREAD);
    tty.c_cflag &= ~static_cast<tcflag_t>(CSTOPB);   // 1 stop bit
    tty.c_cflag &= ~static_cast<tcflag_t>(PARENB);   // no parity
    tty.c_cflag &= ~static_cast<tcflag_t>(CRTSCTS);  // no hardware flow control
    tty.c_cc[VMIN] = 0;
    tty.c_cc[VTIME] = 0;

    if (::tcsetattr(fd_, TCSANOW, &tty) != 0) {
      error = std::strerror(errno);
      close();
      return false;
    }
    ::tcflush(fd_, TCIOFLUSH);
    return true;
  }

  void close()
  {
    if (fd_ >= 0) {
      ::close(fd_);
      fd_ = -1;
    }
  }

  bool isOpen() const {return fd_ >= 0;}

  std::vector<uint8_t> transact(
    const std::vector<uint8_t> & request, uint8_t expect_addr, uint8_t expect_func,
    size_t expect_len, std::vector<uint8_t> * raw_out = nullptr)
  {
    std::vector<uint8_t> buffer;
    if (fd_ < 0) {
      return {};
    }

    ::tcflush(fd_, TCIFLUSH);
    if (::write(fd_, request.data(), request.size()) < 0) {
      return {};
    }
    ::tcdrain(fd_);

    const auto deadline = std::chrono::steady_clock::now() +
      std::chrono::duration_cast<std::chrono::steady_clock::duration>(
      std::chrono::duration<double>(reply_timeout_));

    uint8_t chunk[64];
    std::optional<std::vector<uint8_t>> frame;
    while (!frame) {
      const auto remaining = deadline - std::chrono::steady_clock::now();
      if (remaining <= std::chrono::steady_clock::duration::zero()) {
        break;
      }
      const auto usec = std::chrono::duration_cast<std::chrono::microseconds>(remaining).count();

      timeval tv{static_cast<time_t>(usec / 1000000), static_cast<suseconds_t>(usec % 1000000)};
      fd_set rfds;
      FD_ZERO(&rfds);
      FD_SET(fd_, &rfds);
      if (::select(fd_ + 1, &rfds, nullptr, nullptr, &tv) <= 0) {
        break;
      }

      const ssize_t n = ::read(fd_, chunk, sizeof(chunk));
      if (n <= 0) {
        break;
      }
      buffer.insert(buffer.end(), chunk, chunk + n);
      frame = findFrame(buffer, expect_addr, expect_func, expect_len);
    }

    if (raw_out != nullptr) {
      *raw_out = buffer;
    }
    return frame ? *frame : std::vector<uint8_t>{};
  }

private:
  int fd_{-1};
  double reply_timeout_{0.05};
};

}  // namespace

class Zlac706DiffDrive : public rclcpp::Node
{
public:
  Zlac706DiffDrive()
  : Node("zlac706_diffdrive_node")
  {
    const auto port = declare_parameter<std::string>("port", "/dev/ttyUSB0");
    const int baud = static_cast<int>(declare_parameter<int64_t>("baudrate", 38400));
    wheel_radius_ = declare_parameter<double>("wheel_radius", 0.0508);
    wheel_sep_ = declare_parameter<double>("wheel_separation", 0.23);
    max_rpm_ = declare_parameter<double>("max_rpm", 200.0);
    // 驅動器需要幀間 >=30ms，再加上回應延遲，50ms 一筆交易才穩
    const auto transaction_period = declare_parameter<double>("transaction_period", 0.03);
    invert_left_ = declare_parameter<bool>("invert_left", false);
    invert_right_ = declare_parameter<bool>("invert_right", true);
    cmd_timeout_ = declare_parameter<double>("cmd_timeout", 0.5);
    const auto reply_timeout = declare_parameter<double>("reply_timeout", 0.05);
    debug_serial_ = declare_parameter<bool>("debug_serial", false);

    std::string error;
    if (serial_.open(port, baud, reply_timeout, error)) {
      RCLCPP_INFO(get_logger(), "序列埠已開啟: %s @ %dbps", port.c_str(), baud);
      enableDriver();
    } else {
      RCLCPP_ERROR(get_logger(), "無法開啟序列埠 %s", error.c_str());
    }

    cmd_sub_ = create_subscription<geometry_msgs::msg::Twist>(
      "/cmd_vel", 10, std::bind(&Zlac706DiffDrive::cmdVelCb, this, std::placeholders::_1));
    odom_pub_ = create_publisher<nav_msgs::msg::Odometry>("/odom", 10);
    tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);

    last_time_ = now();
    last_cmd_time_ = std::chrono::steady_clock::now();

    timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::duration<double>(transaction_period)),
      std::bind(&Zlac706DiffDrive::stepCb, this));
  }

  ~Zlac706DiffDrive() override
  {
    writeWheel(ADDR_L, 0.0, invert_left_);
    std::this_thread::sleep_for(std::chrono::milliseconds(40));
    writeWheel(ADDR_R, 0.0, invert_right_);
    serial_.close();
  }

private:
  void enableDriver()
  {
    serial_.transact(ENABLE_ADDR_01, 0x01, 0x06, WRITE_REPLY_LEN);
    std::this_thread::sleep_for(std::chrono::milliseconds(40));
    serial_.transact(ENABLE_ADDR_02, 0x02, 0x06, WRITE_REPLY_LEN);
    std::this_thread::sleep_for(std::chrono::milliseconds(40));
  }

  void writeWheel(uint8_t addr, double rpm, bool invert)
  {
    if (!serial_.isOpen()) {
      return;
    }
    rpm = std::clamp(rpm, -max_rpm_, max_rpm_);
    if (invert) {
      rpm = -rpm;
    }
    serial_.transact(buildSpeedCmd(addr, rpm), addr, 0x06, WRITE_REPLY_LEN);
  }

  std::optional<double> readWheel(uint8_t addr, bool invert)
  {
    if (!serial_.isOpen()) {
      return std::nullopt;
    }
    std::vector<uint8_t> raw;
    const auto rpm_raw = parseSpeedReply(
      serial_.transact(
        readStrpmCmd(addr), addr, 0x03, READ_REPLY_LEN, debug_serial_ ? &raw : nullptr));
    if (!rpm_raw) {
      if (debug_serial_) {
        RCLCPP_WARN(
          get_logger(), "讀取轉速回授失敗 addr=0x%02X | raw: %s", addr, toHex(raw).c_str());
      } else {
        RCLCPP_WARN_THROTTLE(
          get_logger(), *get_clock(), 2000,
          "讀取轉速回授失敗（加 -p debug_serial:=true 可看原始位元組）");
      }
      return std::nullopt;
    }
    const double rpm = rpmFromRaw(*rpm_raw);
    return invert ? -rpm : rpm;
  }

  void cmdVelCb(const geometry_msgs::msg::Twist::SharedPtr msg)
  {
    last_cmd_time_ = std::chrono::steady_clock::now();
    const double v = msg->linear.x;
    const double w = msg->angular.z;

    const double v_left = v - (w * wheel_sep_ / 2.0);
    const double v_right = v + (w * wheel_sep_ / 2.0);

    target_rpm_l_ = (v_left / (2.0 * M_PI * wheel_radius_)) * 60.0;
    target_rpm_r_ = (v_right / (2.0 * M_PI * wheel_radius_)) * 60.0;
  }

  /// 一個 tick 只做一筆交易，timer 週期就是驅動器需要的幀間間隔。
  void stepCb()
  {
    switch (step_) {
      case 0:
        {
          const std::chrono::duration<double> since_cmd =
            std::chrono::steady_clock::now() - last_cmd_time_;
          if (since_cmd.count() > cmd_timeout_) {
            target_rpm_l_ = 0.0;
            target_rpm_r_ = 0.0;
          }
          writeWheel(ADDR_L, target_rpm_l_, invert_left_);
          break;
        }
      case 1:
        writeWheel(ADDR_R, target_rpm_r_, invert_right_);
        break;
      case 2:
        fb_rpm_l_ = readWheel(ADDR_L, invert_left_);
        break;
      default:
        fb_rpm_r_ = readWheel(ADDR_R, invert_right_);
        break;
    }
    // 每個 tick 都積分，TF 才有 20Hz；只在第 4 步發會掉到 5Hz，SLAM 會查不到 odom。
    updateOdom();
    step_ = (step_ + 1) % 4;
  }

  void updateOdom()
  {
    const rclcpp::Time stamp = now();
    const double dt = (stamp - last_time_).seconds();
    last_time_ = stamp;
    if (dt <= 0.0) {
      return;
    }

    // 回授缺任一輪就當靜止，但仍要發 TF，否則 odom->base_link 斷掉會讓 SLAM/Nav2 全掛。
    double v = 0.0;
    double w = 0.0;
    if (fb_rpm_l_ && fb_rpm_r_) {
      const double v_left = (*fb_rpm_l_ / 60.0) * 2.0 * M_PI * wheel_radius_;
      const double v_right = (*fb_rpm_r_ / 60.0) * 2.0 * M_PI * wheel_radius_;
      v = (v_left + v_right) / 2.0;
      w = (v_right - v_left) / wheel_sep_;
    }

    x_ += v * std::cos(theta_) * dt;
    y_ += v * std::sin(theta_) * dt;
    theta_ += w * dt;

    const double quat_z = std::sin(theta_ / 2.0);
    const double quat_w = std::cos(theta_ / 2.0);

    nav_msgs::msg::Odometry odom;
    odom.header.stamp = stamp;
    odom.header.frame_id = "odom";
    odom.child_frame_id = "base_link";
    odom.pose.pose.position.x = x_;
    odom.pose.pose.position.y = y_;
    odom.pose.pose.orientation.z = quat_z;
    odom.pose.pose.orientation.w = quat_w;
    odom.twist.twist.linear.x = v;
    odom.twist.twist.angular.z = w;
    odom_pub_->publish(odom);

    geometry_msgs::msg::TransformStamped tf;
    tf.header.stamp = stamp;
    tf.header.frame_id = "odom";
    tf.child_frame_id = "base_link";
    tf.transform.translation.x = x_;
    tf.transform.translation.y = y_;
    tf.transform.rotation.z = quat_z;
    tf.transform.rotation.w = quat_w;
    tf_broadcaster_->sendTransform(tf);
  }

  SerialPort serial_;

  double wheel_radius_{};
  double wheel_sep_{};
  double max_rpm_{};
  double cmd_timeout_{};
  bool invert_left_{};
  bool invert_right_{};
  bool debug_serial_{false};

  double target_rpm_l_{0.0};
  double target_rpm_r_{0.0};
  std::optional<double> fb_rpm_l_;
  std::optional<double> fb_rpm_r_;
  int step_{0};
  std::chrono::steady_clock::time_point last_cmd_time_;

  double x_{0.0};
  double y_{0.0};
  double theta_{0.0};
  rclcpp::Time last_time_;

  rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr cmd_sub_;
  rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;
  std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Zlac706DiffDrive>());
  rclcpp::shutdown();
  return 0;
}
